# Introduction Loop
kalau di hampir semua bahasa pemrograman yang lain penulisannya sama seperti komponen loop secara Algoritma pada powerpoint, namun berbeda dengan python kita tidak bisa melakukannya seperti tersebut, pada python ketika ingin melakukan looping penulisan syntaxnya menggunakan ***range()***

In [1]:
# fungsi Range pada Python
range(6) # tidak ada outputnya hanya range 0-6

range(0, 6)

In [2]:
# range pada python merupakan generator pada kasus ini, generator tsb menghasilkan 6 bilangan
list(range(6))

[0, 1, 2, 3, 4, 5]

In [3]:
# outputkan satu persatu
for i in range(6):
    print(i) #akan menghasilkan output lalu ganti baris
    #print(i, end=' ') # jika tidak mau ganti baris 

0
1
2
3
4
5


Fungsi range sytnax umumnya ketika kita hanya mengisi satu angka seperti ***list(range(6))*** hanya manghasilkan dari 0 sampai 6-1.

In [4]:
# range pada python 
# Bentuk umum sytax loop pada python range(start, end, step)
list(range(2,10,2)) #jika menelaah code sebelumnya list(range(6))  maka 6 tersebut hanyalah = 'end'
# 10 pada code diatas tidak di outputkan dikarnakan python mengawali index defaultnya adalah 0

[2, 4, 6, 8]

In [5]:
# Contoh penggunaan 'for' loop untuk mengulangi sebanyak 5 kali:
for i in range(5):
    print("Ini adalah iterasi ke-", i)

Ini adalah iterasi ke- 0
Ini adalah iterasi ke- 1
Ini adalah iterasi ke- 2
Ini adalah iterasi ke- 3
Ini adalah iterasi ke- 4


In [6]:
# Menghasilkan deret angka dari 3 hingga 9 (eksklusif):
for i in range(3, 10):
    print("Ini adalah iterasi ke-", i)

Ini adalah iterasi ke- 3
Ini adalah iterasi ke- 4
Ini adalah iterasi ke- 5
Ini adalah iterasi ke- 6
Ini adalah iterasi ke- 7
Ini adalah iterasi ke- 8
Ini adalah iterasi ke- 9


In [7]:
# Menghasilkan deret angka dari 1 hingga 20 (eksklusif) dengan jarak 2:
for i in range(1, 20, 2):
    print("Ini adalah iterasi ke-", i)

Ini adalah iterasi ke- 1
Ini adalah iterasi ke- 3
Ini adalah iterasi ke- 5
Ini adalah iterasi ke- 7
Ini adalah iterasi ke- 9
Ini adalah iterasi ke- 11
Ini adalah iterasi ke- 13
Ini adalah iterasi ke- 15
Ini adalah iterasi ke- 17
Ini adalah iterasi ke- 19


## Komponen loop pada ppt di aplikasikan pada python

In [8]:
SUM = 0
for i in range(1,101):
    SUM = SUM +i #modifikasi +=
SUM

5050

- Kode diatas digunakan untuk menghitung jumlah (sum) dari angka 1 hingga 100 (inklusif) menggunakan 'for' loop dan fungsi range().

- Variabel SUM diinisialisasi dengan nilai 0. Ini akan digunakan untuk menyimpan jumlah total dari angka 1 hingga 100.

- 'for' loop menggunakan fungsi range() dengan argumen start = 1 (nilai awal), stop = 101 (nilai akhir, eksklusif). Dengan demikian, loop akan berjalan dari angka 1 hingga 100.

- Pada setiap iterasi, nilai i (angka saat ini) ditambahkan ke variabel SUM. Alternatif sintaks dengan menggunakan operator += adalah SUM += i, yang berarti sama dengan SUM = SUM + i.

- Setelah 'for' loop selesai, variabel SUM akan berisi jumlah total dari angka 1 hingga 100. Dalam hal ini, SUM akan dicetak sebagai output karena merupakan variabel terakhir dalam kode.



SUM = 0<br>
for i in range(1, 101):<br>
    SUM = SUM + i<br>

Kode diatas menjumlahkan angka dari 1 hingga 100 secara berurutan. Jadi, pada setiap iterasi, kita menambahkan angka yang sesuai ke dalam variabel SUM. Berikut adalah ilustrasi dari apa yang terjadi pada setiap iterasi: :

Iterasi 1: SUM = 0 + 1 = 1<br>
Iterasi 2: SUM = 1 + 2 = 3<br>
Iterasi 3: SUM = 3 + 3 = 6<br>
...<br>
Iterasi 100: SUM = 4950 + 100 = 5050

dalam hal ini, kita menghitung jumlah total angka dari 1 hingga 100 (inklusif), dan hasilnya adalah 5050. Kode ini bukan mencetak angka 1 hingga 100, tetapi menghitung jumlah dari angka-angka tersebut.

In [9]:
# code diatas sama hal nya dengan
sum(range(1,101))

5050

hal ini lebih efisien secara penulisan code karna pada beberapa literatur menyebutkan flat lebih baik dibandingkan neasted, maka dari itu cobalah untuk menuliskan code lebih efisien 

Refrence : <br> 
https://www.datacamp.com/blog/lessons-from-the-zen-of-python

## Hal yang harus di hindari saat melakukan loop
- index loop tidak boleh diletakan didalam loop bodynya disebelah kiri samadengan

In [10]:
b =  0
# index i harus di letakan di sebelah kanan samadengan
for i in range(5):
    b = b + i
print(b)

10


- Kode diatas menggunakan 'for' loop untuk menjumlahkan angka dari 0 hingga 4. Fungsi range(5) menghasilkan angka dari 0 hingga 4 (eksklusif 5). Berikut adalah ilustrasi dari apa yang terjadi pada setiap iterasi:

Iterasi 1: b = 0 + 0 = 0<br>
Iterasi 2: b = 0 + 1 = 1<br>
Iterasi 3: b = 1 + 2 = 3<br>
Iterasi 4: b = 3 + 3 = 6<br>
Iterasi 5: b = 6 + 4 = 10<br>

Setelah loop selesai, variabel b berisi jumlah total dari angka 0 hingga 4, yaitu 10. Oleh karena itu, saat Anda mencetak variabel b, hasilnya adalah 10.

- Alasan utama mengapa indeks loop tidak boleh diletakkan di dalam loop body di sebelah kiri tanda sama dengan (=) adalah karena akan mengubah nilai indeks loop itu sendiri, yang dapat menyebabkan perilaku loop yang tidak terduga dan sulit untuk dilacak.

Misalkan Anda memiliki kode berikut:

In [11]:
b =  0
# index i harus di letakan di sebelah kanan samadengan
for i in range(5):
    i = 2
    b = b + i
print(b)

10


Walaupun kode ini tampak berfungsi, perlu diingat bahwa kita tidak benar-benar mempengaruhi nilai indeks loop itu sendiri. Pada setiap iterasi, nilai i akan diatur ulang berdasarkan urutan angka yang dihasilkan oleh range(5). Jadi, sebenarnya kita tidak mengubah perilaku loop dengan mengubah nilai i di dalam loop body.

Namun, dalam beberapa kasus, seperti saat menggunakan `while loop`, mengubah nilai indeks loop di dalam loop body dapat menyebabkan loop tak terbatas atau perilaku yang tidak diinginkan. Oleh karena itu, sebaiknya hindari mengubah nilai indeks loop secara eksplisit di dalam loop body.

Sebagai praktik terbaik, Anda harus memperlakukan indeks loop sebagai variabel yang hanya digunakan untuk mengontrol loop dan tidak dimodifikasi di dalam loop body. Jika Anda perlu menghitung nilai yang berbeda berdasarkan indeks loop, buat variabel baru di dalam loop body dan gunakan variabel tersebut untuk menyimpan hasil perhitungan Anda.

## Berikut contoh menggunakan list 
- Dalam contoh ini, kita menggunakan range() dengan argumen start = 0 (indeks awal), stop = len(kendaraan) (panjang list), dan step = 2 (jarak antara indeks). Dengan demikian, kita mengakses elemen list kendaraan dengan indeks 0, 2, dan 4, yang merupakan elemen dengan indeks genap.

In [12]:
# Iterasi melalui urutan (sequence)
kumpulan_kendaraan = ['mobil', 'motor', 'bis', 'truk', 'sepeda']
for kendaraan in kumpulan_kendaraan:
    print(kendaraan)

mobil
motor
bis
truk
sepeda


In [13]:
kumpulan_kendaraan = ['mobil', 'motor', 'bis', 'truk', 'sepeda']
for kendaraan in range(0, len(kumpulan_kendaraan), 2):
    print(kumpulan_kendaraan[kendaraan])

mobil
bis
sepeda


Untuk mengubah hasil output dari kode tersebut menjadi list, Anda perlu membuat list kosong terlebih dahulu dan menambahkan elemen ke dalam list tersebut menggunakan metode append(). Berikut adalah contoh kode yang diubah:

In [14]:
# menjadikan output kedalam list
kendaraan = ['mobil', 'motor', 'bis', 'truk', 'sepeda']
hasil = []

for i in range(0, len(kendaraan), 2):
    hasil.append(kendaraan[i])

print(hasil)

['mobil', 'bis', 'sepeda']


Dalam kode ini, kita membuat list kosong bernama hasil. Kemudian, pada setiap iterasi, kita menambahkan elemen dengan indeks genap dari list kendaraan ke dalam list hasil menggunakan metode append(). Setelah perulangan selesai, kita mencetak list hasil yang berisi elemen dengan indeks genap dari list kendaraan.

### Contoh menggunakan enumerate dalam list

In [15]:
#ini biasanya digunakan apabila kita membutuhkan indexnya untuk merubah nilai di dalam listnya
kumpulan_kendaraan = ['mobil', 'motor', 'bis', 'truk', 'sepeda']
for i,a in enumerate(kumpulan_kendaraan):
    print(i,a)

0 mobil
1 motor
2 bis
3 truk
4 sepeda


## Melakukan Penjumlahan seluruh elemen pada sebuah matrix

In [16]:
A = [[1,2,3], [4,5,6], [7,8,9], [4,5,3]] 
SUM = 0
for i in range(4): # untuk setiap baris
    for j in range(3): # untuk setiap kolom
        SUM = SUM + A[i][j] # jumlahkanlah entry" nya
SUM

57

List A adalah list 2 dimensi dengan 4 baris dan 3 kolom

Variabel SUM diinisialisasi dengan nilai 0. Ini akan digunakan untuk menyimpan jumlah total dari semua elemen dalam list A.

Loop for luar akan berjalan dari 0 hingga 3 (eksklusif 4), yang mewakili indeks baris dari list A. Variabel i akan mengambil nilai 0, 1, 2, dan 3.

Loop for dalam akan berjalan dari 0 hingga 2 (eksklusif 3), yang mewakili indeks kolom dari list A. Variabel j akan mengambil nilai 0, 1, dan 2.

Pada setiap iterasi, elemen list A dengan indeks baris i dan indeks kolom j akan ditambahkan ke variabel SUM. Artinya, A[i][j] adalah elemen yang sedang kita proses, dan kita menambahkannya ke SUM: SUM = SUM + A[i][j].

Setelah kedua loop for selesai, variabel SUM akan berisi jumlah total dari semua elemen dalam list A. Dalam hal ini, SUM akan dicetak sebagai output karena merupakan variabel terakhir dalam kode.

In [17]:
# Cara alternatif Kelebihannya adalah cara ini tidak bergantung pada ukuran A
A = [[1,2,3], [4,5,6], [7,8,9], [4,5,3]] 
SUM = 0
for b in A: # untuk setiap baris di matrix
    for x in b: # untuk setiap elemen di baris tersebut
        SUM = SUM + x # jumlahkanlah entry" nya
SUM

57

In [18]:
# Alternatif lebih baik nya melakukan combine perintah "sum" python dan list comperhension
# ini ialah alasan utama  menulis variable SUM dengan huruf kapital
# karena "sum" adalah nama fungsi built-in pada python namespace
A = [[1,2,3], [4,5,6], [7,8,9], [4,5,3]] 

SUM = sum([sum(x) for x in A])
SUM

57

Kode diatas menggunakan list comprehension yang menghasilkan list baru yang berisi jumlah elemen pada setiap baris dalam list A. List comprehension ini akan menciptakan list [sum(x) for x in A], yang akan menghasilkan [6, 15, 24, 12]. Dalam proses ini, kita menggunakan fungsi bawaan sum() untuk menjumlahkan elemen pada setiap baris.

Kemudian, kita menggunakan fungsi bawaan sum() lagi untuk menjumlahkan elemen-elemen dalam list yang dihasilkan oleh list comprehension. Hasilnya adalah 6 + 15 + 24 + 12 = 57.

Nilai ini disimpan dalam variabel SUM. Dalam hal ini, SUM akan dicetak sebagai output karena merupakan variabel terakhir dalam kode.

Kode ini lebih efisien daripada kode sebelumnya karena:

- Kode ini lebih singkat dan lebih mudah dibaca, yang membuatnya lebih mudah untuk dipahami dan dikelola.
- Kode ini menggunakan fungsi bawaan Python sum(), yang dioptimalkan dan umumnya lebih cepat daripada menggunakan loop manual untuk menjumlahkan elemen.

# kapan biasanya for loop dilakukan?

For loop digunakan dalam berbagai situasi di mana Anda perlu mengulangi blok kode untuk setiap elemen dalam suatu urutan atau menjalankan blok kode sejumlah kali tertentu. Selalu ingat untuk menggunakan for loop yang paling sesuai dengan kebutuhan Anda untuk menjaga kode tetap efisien dan mudah dibaca.

# While loop
'while' loop Berbeda dengan 'for' loop yang umumnya digunakan untuk jumlah iterasi yang sudah ditentukan, 'while' loop lebih fleksibel dan cocok untuk situasi di mana kita tidak tahu pasti berapa kali perulangan harus dilakukan.

In [20]:
# contoh looping while di python
i = 0
while i < 9:
    print("Ini adalah iterasi ke-", i)
    i += 1

Ini adalah iterasi ke- 0
Ini adalah iterasi ke- 1
Ini adalah iterasi ke- 2
Ini adalah iterasi ke- 3
Ini adalah iterasi ke- 4
Ini adalah iterasi ke- 5
Ini adalah iterasi ke- 6
Ini adalah iterasi ke- 7
Ini adalah iterasi ke- 8


## Break & Continue
'break' dan 'continue'
Dalam loop, kita bisa menggunakan perintah 'break' untuk keluar dari loop sebelum kondisi menjadi False atau sebelum iterasi selesai. Sementara itu, perintah 'continue' digunakan untuk melewati sisa kode di dalam loop dan langsung melanjutkan ke iterasi berikutnya.

In [21]:
#Contoh penggunaan 'break':
for i in range(10):
    if i == 5:
        break
    print("Ini adalah iterasi ke-", i)

Ini adalah iterasi ke- 0
Ini adalah iterasi ke- 1
Ini adalah iterasi ke- 2
Ini adalah iterasi ke- 3
Ini adalah iterasi ke- 4


In [22]:
#Contoh penggunaan 'continue':
for i in range(10):
    if i % 2 == 0:
        continue
    print("Ini adalah angka ganjil:", i)

Ini adalah angka ganjil: 1
Ini adalah angka ganjil: 3
Ini adalah angka ganjil: 5
Ini adalah angka ganjil: 7
Ini adalah angka ganjil: 9


# For vs While

In [23]:
# contoh looping while di python
n=7
while n>0:
    print(n, end=',')
    n=n-1

7,6,5,4,3,2,1,

Hal diatas merupakan code yang tidak efektif dan buruk, hal tersebut dikarnakan kita sudah mengetahui pasti berapa kali perulangan harus dilakukan, maka gunakanlah For seperti code dibawah.

In [24]:
# For vs While
for i in range(1,8):
    print(8-i, end=', ')

7, 6, 5, 4, 3, 2, 1, 

# While for Domain Constrain
- Dalam Statistika/Data Science/Database while digunakan untuk mencegah noise pada data collection.
- Misal mencegah pengisian nilai lain selain Gender Pria/Wanita pada variable "Jenis Kelamin"

In [25]:
# Contoh
gender = ['pria', 'wanita']
sex=''
while sex not in gender:
    sex = input('Apakah jenis kelamin anda = ')
print('Arigato')

Apakah jenis kelamin anda = laki
Apakah jenis kelamin anda = pria
Arigato


# Try Except
- Secara umum sebuah program sebaiknya harus memiliki "precaution", just in case input yang tidak di inginkan masuk ke dalam sistem dan mengakibatkan error.
- Istilah formal untuk hal ini adalah "Domain Constrain"
- Pada dasarnya hampir semua bahasa pemrograman dan database mampu melakukan domain constrain.

In [27]:
try:
    number = int(input("Masukkan angka: "))
    print("Angka yang dimasukkan adalah:", number)
except ValueError:
    print("Input yang diberikan bukan angka. Silahkan masukkan angka.")

Masukkan angka: 90
Angka yang dimasukkan adalah: 90


Dalam contoh ini, jika pengguna memasukkan input yang bukan angka (seperti teks), int(input()) akan menghasilkan pengecualian ValueError. Dengan menggunakan blok try dan except, kita dapat menangani kesalahan ini dengan menampilkan pesan yang informatif alih-alih menghentikan eksekusi program dengan pesan kesalahan.

## Catch Error 

In [30]:
try:
    n = int(input("please input any integrer = "))
    print("Arigatou")
except Exception as error_:
    print('Error : ', error_)

please input any integrer = cr7
Error :  invalid literal for int() with base 10: 'cr7'


jika masuk kedalam production level output sebelumnya bagus, karna kemudian bagi pengguna informatif, namun bagi programer perlu melihat "Error :  invalid literal for int() with base 10: 'asd'", nah biasanya pada production level, error ini disimpan kedalam error log pada database, ada database tertentu untuk log analysis, dan kemudian dapat di perbaiki karna ada log analysisnya.

## bagaimana jika tidak menggunakan try except?

In [31]:
n = int(input("please input any integrer = "))
print("Arigatou")

please input any integrer = cr7


ValueError: invalid literal for int() with base 10: 'cr7'

bayangkan program pada production level di tokopedia grab gojek, kalaupun ada error, oke error akan disimpan untuk kemudian diperbaiki nanti, dan kemudian programnya tetap jalan, sehingga tetap bisa di gunakan, ***error catching*** pada production level menjadi penting sekali, karna kalu error ini tidak kita tangkap kemudian bisa sitemnya down dan tidak bisa dipakai, bayangkan jika gojek tiba" ada suatu error atau bug di programnya yang kemudian akhiornya membuat astu aplikasi berhenti semua.

# Function
Mendefinisikan Fungsi
Untuk membuat fungsi dalam Python, gunakan kata kunci def diikuti oleh nama fungsi dan tanda kurung yang berisi parameter (jika ada). Kemudian, tambahkan titik dua (:) dan blok kode yang akan dieksekusi ketika fungsi dipanggil.

<img src="Define_a_function.png" alt="Define_a_function" width="650" height="500">
<!-- ![Alt text](./Define_a_function.png) -->

In [32]:
def addtwo(a,b):
    c = a+b
    return c
addtwo(4,6)

10

## Memanggil Fungsi
Untuk memanggil fungsi, cukup gunakan nama fungsi diikuti oleh tanda kurung yang berisi argumen yang sesuai dengan parameter yang didefinisikan dalam fungsi.

In [33]:
def greet():
    print("Hello, World!")
    
greet()  # Output: Hello, World!

Hello, World!


In [34]:
def add(a, b):
    return a + b
add(3, 5)  # result = 8

8

## Parameter dan Argumen
Parameter adalah variabel yang digunakan dalam definisi fungsi, sedangkan argumen adalah nilai yang diberikan saat memanggil fungsi. Parameter didefinisikan dalam tanda kurung setelah nama fungsi, dan argumen ditempatkan dalam tanda kurung saat memanggil fungsi.

In [35]:
def greet(name):
    print(f"Hello, {name}!")

greet("Alice") 

Hello, Alice!


# Contoh kasus

In [36]:
# Fungsi di Python cukup sederhana dan 
# parameter fungsi adalah variabel to reference
# Sebaiknya diberikan multiple lines of metaFunction (keterangan tentang fungsi)
def calculate_flow_rate(volume, time):
    """
    Menghitung debit aliran berdasarkan volume air yang mengalir dan waktu.

    Args:
    volume (float): Volume air yang mengalir dalam liter.
    time (float): Waktu dalam detik.

    Returns:
    float: Debit aliran dalam liter/detik.
    """
    flow_rate = volume / time
    return flow_rate

In [37]:
volume = 100  # Volume air dalam liter
time = 10     # Waktu dalam detik

flow_rate = calculate_flow_rate(volume, time)
print(f"Debit aliran: {flow_rate} liter/detik")  # Output: Debit aliran: 10.0 liter/detik

Debit aliran: 10.0 liter/detik


In [38]:
# Kita bisa melakukan ini kalau kita membuat fungsi dengan "estetika" Python yang benar
help(calculate_flow_rate)

Help on function calculate_flow_rate in module __main__:

calculate_flow_rate(volume, time)
    Menghitung debit aliran berdasarkan volume air yang mengalir dan waktu.
    
    Args:
    volume (float): Volume air yang mengalir dalam liter.
    time (float): Waktu dalam detik.
    
    Returns:
    float: Debit aliran dalam liter/detik.



## Pengaplikasian function pada dataframe menggunakan for loop

<<>>

<!-- Kode ini menghitung debit aliran untuk setiap baris dalam DataFrame `flow_data` dan menyimpan hasil perhitungan dalam kolom baru bernama "flow_rate". Berikut adalah penjelasan langkah demi langkah dari kode tersebut:

1. `flow_data["flow_rate"] = 0`: Ini membuat kolom baru bernama "flow_rate" dalam DataFrame `flow_data` dan menginisialisasi semua nilainya menjadi 0.

2. `for index, row in flow_data.iterrows():`: Menggunakan loop for untuk mengiterasi setiap baris dalam DataFrame `flow_data`. Fungsi `iterrows()` mengembalikan pasangan nilai indeks dan baris dari DataFrame.

3. `volume = row["volume_liters"]`: Mengakses nilai kolom "volume_liters" untuk baris saat ini dan menyimpannya dalam variabel `volume`.

4. `time = row["time_seconds"]`: Mengakses nilai kolom "time_seconds" untuk baris saat ini dan menyimpannya dalam variabel `time`.

5. `flow_rate = calculate_flow_rate(volume, time)`: Memanggil fungsi `calculate_flow_rate` dengan argumen `volume` dan `time` untuk menghitung debit aliran. Hasil perhitungan disimpan dalam variabel `flow_rate`.

6. `flow_data.at[index, "flow_rate"] = flow_rate`: Menggunakan metode `at` untuk mengatur nilai kolom "flow_rate" pada indeks yang sesuai dalam DataFrame `flow_data` dengan hasil perhitungan yang disimpan dalam variabel `flow_rate`.

7. Setelah loop for selesai, `flow_data` akan berisi kolom "flow_rate" yang telah diisi dengan hasil perhitungan debit aliran untuk setiap baris.

Output akhir dari DataFrame `flow_data` setelah menjalankan kode ini akan menampilkan kolom baru "flow_rate" dengan nilai-nilai debit aliran yang telah dihitung untuk setiap baris:

```
   pipe_id  volume_liters  time_seconds  flow_rate
0        1            250            25       10.0
1        2            300            20       15.0
2        3            150            30        5.0
3        4            500            40       12.5
``` -->

In [39]:
import pandas as pd

data = {
    "pipe_id": [1, 2, 3, 4],
    "volume_liters": [250, 300, 150, 500],
    "time_seconds": [25, 20, 30, 40],
}

flow_data = pd.DataFrame(data)
flow_data

pipe_id  volume_liters  time_seconds
0        1            250            25
1        2            300            20
2        3            150            30
3        4            500            40

In [40]:
# Buat kolom baru untuk menyimpan hasil perhitungan debit aliran
flow_data["flow_rate"] = 0

# Menggunakan loop for untuk menghitung debit aliran untuk setiap baris dalam DataFrame
for index, row in flow_data.iterrows():
    volume = row["volume_liters"]
    time = row["time_seconds"]
    flow_rate = calculate_flow_rate(volume, time)
    flow_data.at[index, "flow_rate"] = flow_rate

flow_data

pipe_id  volume_liters  time_seconds  flow_rate
0        1            250            25       10.0
1        2            300            20       15.0
2        3            150            30        5.0
3        4            500            40       12.5

## Pengaplikasian function pada dataframe menggunakan lambda function

<<>>

<!-- Kode ini menghitung debit aliran untuk setiap baris dalam DataFrame `flow_data` menggunakan fungsi `apply()` dan fungsi lambda. Hasil perhitungan disimpan dalam kolom baru bernama "flow_rate". Berikut adalah penjelasan langkah demi langkah dari kode tersebut:

1. `flow_data.apply(...)`: Metode `apply()` pada objek DataFrame digunakan untuk menerapkan suatu fungsi pada seluruh DataFrame atau pada sumbu tertentu (baris atau kolom). Dalam kasus ini, kita ingin menerapkan fungsi lambda pada setiap baris DataFrame, jadi kita menggunakan argumen `axis=1`.

2. `lambda row: ...`: Fungsi lambda didefinisikan untuk mengambil satu argumen, yaitu `row`. Fungsi lambda akan dipanggil dengan setiap baris DataFrame sebagai argumen.

3. `calculate_flow_rate(row["volume_liters"], row["time_seconds"])`: Di dalam fungsi lambda, kita memanggil fungsi `calculate_flow_rate()` dengan argumen yang diambil dari kolom "volume_liters" dan "time_seconds" pada baris saat ini. Hasil perhitungan debit aliran akan menjadi nilai yang dikembalikan oleh fungsi lambda.

4. `flow_data["flow_rate"] = ...`: Hasil perhitungan debit aliran untuk setiap baris akan disimpan dalam kolom baru bernama "flow_rate" pada DataFrame `flow_data`.

Setelah menjalankan kode ini, DataFrame `flow_data` akan berisi kolom "flow_rate" yang telah diisi dengan hasil perhitungan debit aliran untuk setiap baris:

```
   pipe_id  volume_liters  time_seconds  flow_rate
0        1            250            25       10.0
1        2            300            20       15.0
2        3            150            30        5.0
3        4            500            40       12.5
```

Dengan pendekatan ini, kita dapat menghitung debit aliran dalam satu baris kode menggunakan fungsi `apply()` dan fungsi lambda, sehingga kode menjadi lebih ringkas dan mudah dibaca. -->

In [11]:
flow_data["flow_rate"] = flow_data.apply(lambda row: calculate_flow_rate(row["volume_liters"], row["time_seconds"]), axis=1)
flow_data

pipe_id  volume_liters  time_seconds  flow_rate
0        1            250            25       10.0
1        2            300            20       15.0
2        3            150            30        5.0
3        4            500            40       12.5

# Soal Function

In [60]:
import pandas as pd

data = {
    "particle_id": [1, 2, 3, 4],
    "distance_meters": [100, 200, 150, 250],
    "time_seconds": [10, 20, 15, 25],
    "initial_speed_mps": [0, 5, 10, 15],
    "final_speed_mps": [20, 15, 25, 20],
    "mass_kg": [5, 10, 7, 12],
}

particle_data = pd.DataFrame(data)
particle_data

particle_id  distance_meters  time_seconds  initial_speed_mps  \
0            1              100            10                  0   
1            2              200            20                  5   
2            3              150            15                 10   
3            4              250            25                 15   

   final_speed_mps  mass_kg  
0               20        5  
1               15       10  
2               25        7  
3               20       12

1) Buatlah fungsi untuk menghitung kecepatan rata-rata partikel menggunakan rumus: ***average_speed = distance / time***. Gunakan DataFrame di atas dan hitunglah kecepatan rata-rata untuk setiap partikel.


2) Buatlah fungsi untuk menghitung percepatan partikel menggunakan rumus: ***acceleration = (final_speed - initial_speed) / time***. Gunakan DataFrame di atas dan hitunglah percepatan untuk setiap partikel.


3) Buatlah fungsi untuk menghitung energi kinetik partikel menggunakan rumus: ***kinetic_energy = 0.5 * mass * (final_speed ^ 2)***. Gunakan DataFrame di atas dan hitunglah energi kinetik untuk setiap partikel.


4) Tambahkan kolom baru pada DataFrame untuk menyimpan hasil perhitungan kecepatan rata-rata, percepatan, dan energi kinetik. Tampilkan DataFrame yang telah diperbarui.

Dalam soal ini, siswa diharapkan untuk memahami konsep fungsi dalam Python, serta bagaimana menggunakan fungsi tersebut untuk mengolah data dalam DataFrame dan menerapkan rumus fisika instrumental.

In [61]:
#Soal 1
def calculate_average_speed(distance, time):
    average_speed = distance / time
    return average_speed
# Buat kolom baru untuk menyimpan hasil perhitungan debit aliran
particle_data["average_speed"] = 0

# Menggunakan loop for untuk average speed
for index, row in particle_data.iterrows():
    distance = row["distance_meters"]
    time = row["time_seconds"]
    average_speed = calculate_average_speed(distance, time)
    particle_data.at[index, "average_speed"] = average_speed

particle_data

particle_id  distance_meters  time_seconds  initial_speed_mps  \
0            1              100            10                  0   
1            2              200            20                  5   
2            3              150            15                 10   
3            4              250            25                 15   

   final_speed_mps  mass_kg  average_speed  
0               20        5             10  
1               15       10             10  
2               25        7             10  
3               20       12             10

In [62]:
#Soal 2
def calculate_delta_speed(final, initial):
    delta_speed = (final-initial)
    return delta_speed

# Buat kolom baru untuk menyimpan hasil delta speed
particle_data["delta_speed"] = 0

# Menggunakan loop for untuk delta speed
for index, row in particle_data.iterrows():
    final = row["final_speed_mps"]
    initial = row["initial_speed_mps"]
    delta_speed = calculate_delta_speed(final, initial)
    particle_data.at[index, "delta_speed"] = delta_speed

    
def calculate_acceleration(delta_speed, time):
    acceleration = (delta_speed / time)
    return acceleration

# Buat kolom baru untuk menyimpan hasil perhitungan acceleration
particle_data["acceleration"] = 0

# Menggunakan loop for untuk acceleration
for index, row in particle_data.iterrows():
    delta_speed = row["delta_speed"]
    time = row["time_seconds"]
    acceleration = calculate_acceleration(delta_speed, time)
    particle_data.at[index, "acceleration"] = acceleration    
    
particle_data

particle_id  distance_meters  time_seconds  initial_speed_mps  \
0            1              100            10                  0   
1            2              200            20                  5   
2            3              150            15                 10   
3            4              250            25                 15   

   final_speed_mps  mass_kg  average_speed  delta_speed  acceleration  
0               20        5             10           20           2.0  
1               15       10             10           10           0.5  
2               25        7             10           15           1.0  
3               20       12             10            5           0.2

In [63]:
#Soal 3
def calculate_kinetic_energy (mass, final_speed):
    kinetic_energy = (0.5 * mass * (final_speed**2))
    return kinetic_energy

# Buat kolom baru untuk menyimpan hasil perhitungan acceleration
particle_data["kinetic_energy"] = 0

# Menggunakan loop for untuk acceleration
for index, row in particle_data.iterrows():
    mass = row["mass_kg"]
    final_speed = row["final_speed_mps"]
    kinetic_energy = calculate_kinetic_energy(mass, final_speed)
    particle_data.at[index, "kinetic_energy"] = kinetic_energy    
    
particle_data

particle_id  distance_meters  time_seconds  initial_speed_mps  \
0            1              100            10                  0   
1            2              200            20                  5   
2            3              150            15                 10   
3            4              250            25                 15   

   final_speed_mps  mass_kg  average_speed  delta_speed  acceleration  \
0               20        5             10           20           2.0   
1               15       10             10           10           0.5   
2               25        7             10           15           1.0   
3               20       12             10            5           0.2   

   kinetic_energy  
0          1000.0  
1          1125.0  
2          2187.5  
3          2400.0

In [79]:
#Soal 4
def average_speed(data):
    df = data.assign(average_speed_mps=(data["distance_meters"] / data["time_seconds"]))
    return df
particle_data = average_speed(particle_data)

def acceleration(data):
    df = data.assign(acceleration_mps2=(data["final_speed_mps"] - data["initial_speed_mps"]) / data["time_seconds"])
    return df
particle_data = acceleration(particle_data)

def kinetic_energy(data):
    df = data.assign(kinetic_energy_joule = (0.5 * data["mass_kg"] * (data["time_seconds"]**2)))
    return df
particle_data = kinetic_energy(particle_data)
particle_data

particle_id  distance_meters  time_seconds  initial_speed_mps  \
0            1              100            10                  0   
1            2              200            20                  5   
2            3              150            15                 10   
3            4              250            25                 15   

   final_speed_mps  mass_kg  average_speed  delta_speed  acceleration  \
0               20        5             10           20           2.0   
1               15       10             10           10           0.5   
2               25        7             10           15           1.0   
3               20       12             10            5           0.2   

   kinetic_energy  acceleration_mps2  kinetic_energy_joule  average_speed_mps  
0          1000.0                2.0                 250.0               10.0  
1          1125.0                0.5                2000.0               10.0  
2          2187.5                1.0                 787.5               10.0  
3          2400.0                0.2                3750.0               10.0

<<>>

<!-- import pandas as pd

data = {
    "particle_id": [1, 2, 3, 4],
    "distance_meters": [100, 200, 150, 250],
    "time_seconds": [10, 20, 15, 25],
    "initial_speed_mps": [0, 5, 10, 15],
    "final_speed_mps": [20, 15, 25, 20],
    "mass_kg": [5, 10, 7, 12],
}

particle_data = pd.DataFrame(data)

def acceleration(data):
    df = data.assign(acceleration_mps2=(data["final_speed_mps"] - data["initial_speed_mps"]) / data["time_seconds"])
    return df

particle_data = acceleration(particle_data)
particle_data 

Menggunakan metode .assign(), Anda dapat menambahkan kolom baru ke DataFrame dengan menghitung percepatan berdasarkan kolom "initial_speed_mps", "final_speed_mps", dan "time_seconds" tanpa menggunakan loop for atau fungsi lambda-->